litgpt/generate/base.py

```diff
@@ -75,8 +75,56 @@ def sample(
 
 def next_token(model: GPT, input_pos: torch.Tensor, x: torch.Tensor, **kwargs: Any) -> torch.Tensor:
     logits = model(x, input_pos)
-    next = sample(logits, **kwargs)
-    return next.to(dtype=x.dtype)
+    _next = sample(logits, **kwargs).to(dtype=torch.int64)
+    return _next
+
+
+def batched_sample(logits: list[torch.Tensor], kwargs: list[dict]) -> list[torch.Tensor]:
+    assert len(logits) == len(kwargs), "logits and kwargs must have the same length."
+    return [sample(l, **sample_args).to(dtype=torch.int64) for sample_args, l in zip(kwargs, logits)]
+
+
+def batched_next_token(model: GPT, input_pos: torch.Tensor, x: list[torch.Tensor], kwargs: Union[dict, list[dict]]) -> list[torch.Tensor]:
+    # Where
+    # input_pos is a tensor of shape [seq_length...]
+    # Each tensor in x is 1d, and represents the context tokens to add to the kvcache.
+
+    # TODO: Take input_pos as a list of tensors.
+    # The desired API is input_pos: Union[torch.Tensor, list[torch.Tensor]].
+    # This means making the rope cache and kvcache forward() work with batches. Currently, they do not.
+    # This is relatively complicated, given the current implementation. It will require some rewriting.
+    # Relevant thread: https://discuss.pytorch.org/t/batched-index-select/9115
+    # We will also need the same with tensor.index_copy_(). These do not work for batches, and the replacement
+    # is somewhat nontrivial. Until then, we can only accept prompts that are all the same length.
+    # After this problem is resolved, there will be another problem. That being, continuous batched prefill.
+    # If you have any ideas on this, let me know. I don't think that padding input_pos is viable.
```

Comment:
As mentioned on slack, I have had good experience with padding input_pos and input with the rightmost entries, but maybe we can just do a v0 with (necessarily) rectangular tensors and keep the padding out of this for now. That would be most useful.